crates/syntax/src/algo.rs

```diff
@@ -108,40 +110,53 @@ pub enum InsertPosition<T> {
 
 pub struct TreeDiff {
     replacements: FxHashMap<SyntaxElement, SyntaxElement>,
+    deletions: Vec<SyntaxElement>,
+    // the vec as well as the indexmap are both here to preserve order
+    insertions:
+        IndexMap<SyntaxElement, Vec<SyntaxElement>, BuildHasherDefault<rustc_hash::FxHasher>>,
```

Comment:
Insertion order is important as insertions might be chained after each other. The `IndexMap` could be replace by `Vec<(SyntaxElement, Vec<SyntaxElement>)` with binary_search insertion on the first tuple part as the key but I'm uncertain what would be better.